### PR TITLE
Prepare for GHC 9.2.

### PR DESCRIPTION
**Pull request by erikd**
## Related Issue

ADP-2679

## Summary

This PR:
- updates dependencies of `cardano-wallet` to prepare for building with GHC 9.2.
- adds basic support for the Conway era, such that existing test suites pass.

This PR does **NOT**:

- add CI support for GHC 9.2.
- add extra test suite coverage that's specific to Conway.

We will add an appropriate CI support for GHC 9.2 in a **_subsequent_** PR.

### CODE REVIEW

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: Suggestion: use `genNonEmpty` and `shrinkNonEmpty` from `Test.QuickCheck.Extra`

```suggestion
  arbitrary = genNonEmpty arbitrary
  shrink = shrinkNonEmpty shrink
```

File: lib/wallet/api/http/Cardano/Wallet/Launch/Cluster.hs
Comment (jonathanknowles) [inline lib/wallet/api/http/Cardano/Wallet/Launch/Cluster.hs]: Just curious, is it possible to write:

```suggestion
                    { Ledger.sgsPools = (ListMap [(poolId, params)])
```

File: lib/wallet/cardano-wallet.cabal
Comment (jonathanknowles) [inline lib/wallet/cardano-wallet.cabal:185]: Finally! 💯 

This means we can now depend upon the hackage version of `strict-containers` (if we need to).

File: lib/wallet/src/Cardano/Api/Gen.hs
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Api/Gen.hs]: Just some minor formatting fixes (our house style favours newline + 4-space-indent instead of vertical alignment).

```suggestion
    let costModelParams = Alonzo.getCostModelParams Plutus.testingCostModelV1
    eCostModel <- Alonzo.mkCostModel
        <$> genPlutusLanguage
        <*> mapM (const $ chooseInteger (0, 5_000)) costModelParams
    case eCostModel of
        Left err -> error $ "genCostModel: " ++ show err
        Right cModel -> return . CostModel $ Alonzo.getCostModelParams cModel
```

File: lib/wallet/src/Cardano/Wallet/Read/Primitive/Tx/Allegra.hs
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Read/Primitive/Tx/Allegra.hs]: This will allow `stylish-haskell` to re-order the imports properly:

```suggestion
    ( (^.) )
import Cardano.Ledger.Core
```

File: lib/wallet/src/Cardano/Wallet/Read/Primitive/Tx/Mary.hs
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Read/Primitive/Tx/Mary.hs]:
```suggestion
    ( (^.) )
import Cardano.Api
```

**Review by jonathanknowles — COMMENTED** [review]
Hi @erikd 

Many thanks for making this PR!

I've had a brief look, and have some initial (small) suggestions.

Will take another look later on, when I get some time.

Thanks again!
Jonathan

File: lib/wallet/api/http/Cardano/Wallet/Launch/Cluster.hs
Comment (erikd) [reply lib/wallet/api/http/Cardano/Wallet/Launch/Cluster.hs]: I find the `fromList` version more obvious.

File: cabal.project
Comment (Anviking) [inline cabal.project:116]: It seems all of this should be removed then:
```haskell
-- -- Different version than from CHaP to include the ReaderT fix.
 -- -- TODO [ADP-2447] This source-repository-package should be removable with next node bump.
 -- source-repository-package
 --   type: git
 --   location: https://github.com/input-output-hk/io-sim
 --   tag: 1cdeed83b4cccf4ed1e656375200c889ecd879c7
 --   --sha256: 1wy33hs5rg2n0haf5wc9s9mrrkdzyp95njkkpgs8g5mnllapdka8
 --   subdir:
 --     io-classes
 --     io-sim
 --     strict-stm

 -- It seems that if we remove this source-repository-package to rely on CHaP,
 -- for some reason, we're unable to retrieve the cardano-node executables in
 -- our nix/haskell.nix file.
```

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]:
```suggestion
```

**Comment by paolino** [comment]
Hi, Thanks for the work in this repo.

I'm fixing stylish haskell / imports issues and hlint in 2 separate additional commits.

But I cannot work with the code actually. 

```
[paolino@nixos:~/iohk/cardano-wallet]$ nix/regenerate.sh 
No changes needed to cabal.project
error: A definition for option `compiler-nix-name' is not of type `string'. Definition values:
       - In `<unknown-file>': null
(use '--show-trace' to show detailed location information)
```
same error for nix develop

Does anyone have a fix?

File: cabal.project
Comment (erikd) [reply cabal.project:116]: I removed the `io-sim` part of that comment block.

**Comment by erikd** [comment]
I pulled all these changes, quashed them down and force pushed.

I also rebased it against master.

**Review by jonathanknowles — DISMISSED** [review]

**Comment by jonathanknowles** [comment]
Hi @erikd 

Since we can't yet build this within the nix environment, this means that we can't yet run the test suite to verify that the tests still pass.

Therefore, we'll have to wait for the nix configuration to be fixed before we can approve and merge this PR.

The work to fix the nix configuration is happening here: https://github.com/input-output-hk/cardano-wallet/pull/3700

**Comment by michaelpj** [comment]
You get the `compiler-nix-name` from the `with-compiler` field of the `cabal.project`, which is commented out here. See https://github.com/input-output-hk/cardano-wallet/blob/e2749fb9daf52a9310dc7fd4df387b09e8cfb0da/nix/overlays/common-lib.nix#L49

This is inherently incompatible with building the project with two GHCs (in Nix at least).

This behaviour was introduced [here](https://github.com/input-output-hk/cardano-wallet/commit/0b307bdc1853db4ede28f047b8fc287da56632ea) and is partially documented in your [contributing docs](https://github.com/input-output-hk/cardano-wallet/blob/e2749fb9daf52a9310dc7fd4df387b09e8cfb0da/docs/contributing/Notes-from-upgrading-GHC-version.md#process).

You probably need to rearrange things to take the compiler name as a parameter if you want your Nix build to work with both. If you're happy to just have the Nix stuff work with GHC 9.X only, just put the `with-compiler` field back in.

**Comment by angerman** [comment]
For now code *must* compile with 8.10 *and* 9.2. We can not retire 8.10 until we have published 9.2 builds that have passed extensive Q/A. Until then we need to have 8.10 as a fallback option.

File: cabal.project
Comment (Anviking) [inline cabal.project:212]: Is this ok to change?

**Review by Anviking — APPROVED** [review]

File: cabal.project
Comment (jonathanknowles) [reply cabal.project:212]: Without this change, I could not get this branch to build (as cabal was not able to satisfy the dependency). I honestly don't know what effect it will have, if any.

**Review by jonathanknowles — APPROVED** [review]

File: cabal.project
Comment (angerman) [reply cabal.project:212]: It just makes cardano-node not depend on systemd for logging. I'd argue depending on systemd is wrong to begin with though. Using it is a different topic but having a hard dependency on it?

All static node builds have this disabled as well. It has no functional impact.

**Pull request merged**
